Fix: Remove Replenishment with "0" Value on "to order"

Humml87 opened the pull request:
Now a row for a "replenishmet" with a value "0" on "To Order" not creating.

As example a 0 can a result from "STOCK_REPLENISH_ADD_CHECKBOX_INCLUDE_DRAFT_ORDER" with active checkbox "&draftchecked".

Commit:
Fix ZeroToOrder in Replenish

stickler-ci committed:
Fixing style errors.

eldy commented on htdocs/product/stock/replenish.php:
```diff
@@ -875,6 +875,8 @@
 			$stocktobuywarehouse = max(max($desiredstockwarehouse, $alertstockwarehouse) - $stockwarehouse, 0); //ordered is already in $stock in virtual mode
 		}
 
+		$toOrder = ((!empty($conf->global->STOCK_ALLOW_ADD_LIMIT_STOCK_BY_WAREHOUSE) && $fk_entrepot > 0) > 0 ? $stocktobuywarehouse : $stocktobuy);
```
(!empty($conf->global->STOCK_ALLOW_ADD_LIMIT_STOCK_BY_WAREHOUSE) && $fk_entrepot > 0) is a boolen.
But you compare it with > 0 
Surely a bug.

Humml87 replied:
It was a c&p fail, sorry. I changed this now to a clean condition.

Commit:
Clear conditions on $toOrder. Cleanup.

Commit:
Merge remote-tracking branch 'origin/develop_Fix_ZeroToOrder_in_Replenish' into develop_Fix_ZeroToOrder_in_Replenish

Humml87 commented:
I make a review on this PR. I think it is better to filter the useless replenishments with the initial SQL query.
Currently we have also the problem that the "Replensishment Counter" not correct. See the following screenshot.
![image](https://user-images.githubusercontent.com/78662388/209811217-6843e687-4115-4ea8-9a60-97dc1c4a5bc9.png)
The query include the services also, if the option "STOCK_SUPPORTS_SERVICES" not set. So the counter is not correct.
![image](https://user-images.githubusercontent.com/78662388/209811577-89086c3e-3197-4801-bf3a-02408fae1ad1.png)

Humml87 committed:
Merge branch 'Dolibarr:develop' into develop_Fix_ZeroToOrder_in_Replenish

Commit:
Clear some functional and logical bugs.
- Also Orders with the State "Validated" and "Accepted" are observed.
- Reduce duplicated code.
- Rename "IncludeDraft" => IncludeDraftToAccepted.
- Remove Replenishment Rows with Zero to order.
- Replenishment counter now correct.

stickler-ci committed:
Fixing style errors.

Humml87 committed:
Merge branch 'Dolibarr:develop' into develop_Fix_ZeroToOrder_in_Replenish

Commit:
Fix SQL Query for Travis.

stickler-ci committed:
Fixing style errors.

eldy commented:
I can't reproduce the trouble. Do you know how to create data to reproduce the bug ?

Humml87 commented:
For the first Problem "0 Value":
Step 1:
![image](https://user-images.githubusercontent.com/78662388/211161511-e13f8115-b7ef-4726-bf1c-a597b6dea25b.png)
Step 2:
![image](https://user-images.githubusercontent.com/78662388/211161688-843a2ab7-f98a-43e5-a20f-e9f1e0e1af57.png)
Step 3:
![image](https://user-images.githubusercontent.com/78662388/211161799-efc3323d-3992-428c-8e2e-7a58756e413b.png)

Humml87 commented:
For the second Problem "Wrong Counter" (option "STOCK_SUPPORTS_SERVICES" **not** set):
Step 1:
![image](https://user-images.githubusercontent.com/78662388/211162174-deae1589-baad-43cd-924a-a26e1231c315.png)
Step 2:
![image](https://user-images.githubusercontent.com/78662388/211162287-a27f0646-e418-4cfe-b4e4-32748dd4c009.png)
Step 3:
![image](https://user-images.githubusercontent.com/78662388/211162336-bb2f65cb-86cf-4125-bcd2-b1beaa7be0df.png)
Step 4:
![image](https://user-images.githubusercontent.com/78662388/211162407-0c8bb9b9-be38-4aeb-a19e-09ffed4d4a93.png)

Humml87 commented:
@eldy two examples to reproduce the troubles.

Humml87 commented:
@eldy can you merge/review? I like to start with the next Issues (#23446) which builds on this PR.
Thank you!

eldy commented:
> For the first Problem "0 Value":
> Step 1:
> ![image](https://user-images.githubusercontent.com/78662388/211161511-e13f8115-b7ef-4726-bf1c-a597b6dea25b.png)
> Step 2:
> ![image](https://user-images.githubusercontent.com/78662388/211161688-843a2ab7-f98a-43e5-a20f-e9f1e0e1af57.png)
> Step 3:
> ![image](https://user-images.githubusercontent.com/78662388/211161799-efc3323d-3992-428c-8e2e-7a58756e413b.png)
> 

This is the expected behaviour to have 0 here. The suggested value to order is based on the "desired stock", not on the "limit for alert". Limit for alert is just to say when we want a warning. Desired stock is to say value to reach for reassort. On your example, value of desired stock is not defined, so suggested amount to order is 0 (no target defined = no suggestion to reach target)

Humml87 commented:
We talk about different situations.
In the shown example, i have set the option **"include also draft orderes"**. So the proposal to replenish the product is not right.
I hope i can show you the problem with a few pictures:
![image](https://user-images.githubusercontent.com/78662388/215473240-3b7365df-195e-491d-81b2-41b0bac56826.png)
![image](https://user-images.githubusercontent.com/78662388/215475949-fd3f4c67-9207-4cdd-b7d3-c312450c3283.png)

![image](https://user-images.githubusercontent.com/78662388/215477553-41bf9c3c-6ff0-48b8-a49c-6c7c9b66fc74.png)

The "Desired Stock" is not important on this. If the option "include also draft orders" is enabled then we should pretend the order is ordered.

> _**"On your example, value of desired stock is not defined, so suggested amount to order is 0 (no target defined = no suggestion to reach target)"**_

=> Currently, if no "Desired Stock" is set, so the replenishment make a proposal to stock up to "Limit for alert". I think this is correct.

Best regards

eldy commented:
The "limit for alert" must be used on for alert.
The "desired stock" is used for reassort and only reassort.
So if Desired stock is not defined, the expected behaviour is to suggest 0. End user has defined "No target" for stock reassort so no suggestion during reassort must be done.
Including draft order or not will and must not change this. This is the user choice to not have reassort target to not include product into reassort process. So event if there is draft order, if "desired stock" is empty, no order will be suggested. 
Including draft order may change the quantity to reach the target if a target has been defined.

I did not succeed in reproducing a use case where this expected behaviour is not working.

Humml87 commented:
> So if Desired stock is not defined, the expected behaviour is to suggest 0. End user has defined "No target" for stock reassort so no suggestion during reassort must be done.

The Problem with replenishment lines with a "0" value is the workflow to create a order look like this:
![image](https://github.com/Dolibarr/dolibarr/assets/78662388/fd3d7f24-e3ee-40ab-8ce9-553f80908574)
The User must always disable all "0" value lines to create an order. And the clarity with many of "0" value lines is not good (we have hundreds of this). I think no User needs a replenishment line with "0".

Best regards Christian

eldy commented:
The message "Some orders haven't been created because of too-lo quantities" is not related with the line to order at 0.
When "to order" is 0 (for example when desired stock for reassort is empty on product), nothiing shoul dbe created for this line.
The message seems related to other line. dolibarr try to create orders but you probably have the option to disallow creation of order if quantity is not enough (In setup of module "Stock")

Humml87 commented:
> This is the expected behaviour to have 0 here.

Hi @eldy, sorry for disturb you for this topic again.
I become some feedback from my purchasing department. please view the following screenshot. We have in this case 199 open replenishment's. Much replenishment's of there are ordered already, but still in the list.
It's really uncomfortable to create orders for the replenishment's, where "To order" is grather then 0.

![image](https://github.com/Dolibarr/dolibarr/assets/78662388/ef082f60-c60e-4a91-836f-6c34ffc1c1c2)
I hope you can understand our problem.

What do you think of the idea to implement a new setting in the module "Stock" to hide lines with "To order" are 0?
![image](https://github.com/Dolibarr/dolibarr/assets/78662388/528367f7-1d46-494a-bde5-fef8c98ab8a9)

Best regards
Christian

eldy commented:
So we need something to reduce the list. There is other reason to have suggested value of 0.
So I think the label of option could be 
"Hide lines with a stock already higher than the expected qty"
So you can filter lines with virtual stock > desired stock.
I don't think this must be into the setup but should be an the replenishment page.
